### PR TITLE
fix: add support for custom properties for braze purchase events

### PR DESCRIPTION
**Pull request by utsabc**
## Description of the change

Adding support for `properties` fields for Purchase events to Braze [ref](https://www.braze.com/docs/api/objects_filters/purchase_object/)

Screenshot 2023-11-29 at 6 06 27 PM

The `properties` field is mapped with all the extra fields perent in each of the `product` elements present in `Order Completed` event.

This change introduces additional data to be sent into Braze

Refer to linear ticket INT-1076 for further details

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/braze/util.js [hunk 1]
```diff
@@ -634,6 +634,10 @@ function getPurchaseObjs(message) {
       parseInt(quantity, 10),
       timestamp,
     );
+    const extraProperties = _.omit(product, ['product_id', 'sku', 'price', 'quantity', 'currency']);
```
Comment (koladilip) [inline src/v0/destinations/braze/util.js:637]: should be enable this via destination config to avoid impacting all customer, we don't know if everyone needs this or not

Comment (utsabc) [reply src/v0/destinations/braze/util.js:637]: Its not a good `product UX` that for a small feature like this we add a config in UI?

**Review by ItsSudip — DISMISSED** [review]

File: src/v0/destinations/braze/util.js (hunk 1 above)
Comment (utsabc) [reply src/v0/destinations/braze/util.js:637]: We are handling this behind a destination config

Comment (koladilip) [inline src/v0/destinations/braze/util.js:637]: if (Config.sendPurchaseEventWithExtraProperties) {
do omit logic then add to purchase object

doing omit without this config flag is useless right?
}

File: src/v0/destinations/braze/util.js [hunk 2]
```diff
@@ -539,7 +539,7 @@ function addMandatoryPurchaseProperties(productId, price, currencyCode, quantity
   };
 }
 
-function getPurchaseObjs(message) {
+function getPurchaseObjs(message, Config) {
```
Comment (krishna2020) [inline src/v0/destinations/braze/util.js:542]:
```suggestion
function getPurchaseObjs(message, config) {
```

File: src/v0/destinations/braze/util.js (hunk 1 above)
Comment (krishna2020) [inline src/v0/destinations/braze/util.js:637]: can we define this list at one place & use it for mappings & for omitting here ?
so that any new field mapping requirement, change at one place would be sufficient.
this can go as part of different PR too

Comment (utsabc) [reply src/v0/destinations/braze/util.js:637]: I am also checking the `extraProperties` length along with config hence I am extracting it outside

File: src/v0/destinations/braze/util.js (hunk 2 above)
Comment (utsabc) [reply src/v0/destinations/braze/util.js]: Addressed

File: src/v0/destinations/braze/util.js (hunk 1 above)
Comment (utsabc) [reply src/v0/destinations/braze/util.js]: Addressed

Comment (koladilip) [reply src/v0/destinations/braze/util.js]: but omitting operation itself is not need right if condition is not met?

**Review by ItsSudip — APPROVED** [review]
